### PR TITLE
Fix #21325, use special version of eltype

### LINKED ISSUE #21325: Broadcast eltype error
While working on #21290, I discovered an error in how `_broadcast_eltype` determines the element type. The error manifests when the return value of a function on a broadcast-scalar argument is distinct from that of the argument's `eltype`.

```julia
julia> foo(x::Char, y::Int) = 0
foo (generic function with 1 method)

julia> foo(x::String, y::Int) = "hello"
foo (generic function with 2 methods)

julia> Base.Broadcast.broadcast(foo, "x", [1, 2, 3])
ERROR: MethodError: Cannot `convert` an object of type String to an object of type Int64
This may have arisen from a call to the constructor Int64(...),
since type constructors fall back to convert methods.
Stacktrace:
 [1] setindex! at ./multidimensional.jl:247 [inlined]
 [2] macro expansion at ./broadcast.jl:152 [inlined]
 [3] macro expansion at ./simdloop.jl:72 [inlined]
 [4] macro expansion at ./broadcast.jl:145 [inlined]
 [5] _broadcast!(::#foo, ::Array{Int64,1}, ::Tuple{Tuple{},Tuple{Bool}}, ::Tuple{Tuple{},Tuple{Int64}}, ::String, ::Tuple{Array{Int64,1}}, ::Type{Val{1}}, ::CartesianRange{CartesianIndex{1}}) at ./broadcast.jl:138
 [6] broadcast_t(::Function, ::Type{T} where T, ::Tuple{Base.OneTo{Int64}}, ::CartesianRange{CartesianIndex{1}}, ::String, ::Array{Int64,1}) at ./broadcast.jl:267
 [7] broadcast_c at ./broadcast.jl:293 [inlined]
 [8] broadcast(::Function, ::String, ::Array{Int64,1}) at ./broadcast.jl:415
```

I should have a fix, which I'll post later today or tomorrow.

### PR DESCRIPTION
**Pull request by TotalVerb**
Currently `_broadcast_eltype` calls inference but can often do it on the wrong types, which is not good. Calling `eltype` to get those types was incorrect because `broadcast` treats more things as scalars than what `eltype` does.

This PR replaces `eltypestuple` with a more general `maptuple` and adds two `_*_eltype` functions. We also combine `Base.eltypes` and `Base.eltypestuple` into the `maptuple` function.

A further nice consequence is that in many cases where this bug was nearly invisible before (because inference failed on one of the arguments), the return of `broadcast` can now be inferred, and performance is improved. (I don't know if any such benchmarks are in BaseBenchmarks, but I plan to add some.)

### CODE REVIEW

**Comment by TotalVerb** [comment]
cc @Sacha0 
cc @pabloferz

File: base/sparse/higherorderfns.jl [hunk 1]
```diff
@@ -85,7 +85,7 @@ function _noshapecheck_map{Tf,N}(f::Tf, A::SparseVecOrMat, Bs::Vararg{SparseVecO
     fofzeros = f(_zeros_eltypes(A, Bs...)...)
     fpreszeros = _iszero(fofzeros)
     maxnnzC = fpreszeros ? min(length(A), _sumnnzs(A, Bs...)) : length(A)
-    entrytypeC = Base.Broadcast._broadcast_eltype(f, A, Bs...)
+    entrytypeC = Base.Broadcast._broadcast_eltype(A, Bs...)
```
Comment (Sacha0) [inline base/sparse/higherorderfns.jl]: Function argument removed accidentally?

**Comment by Sacha0** [comment]
@TotalVerb, great catch! Will review more thoroughly later this weekend. Thanks and best!

File: base/broadcast.jl [hunk 2]
```diff
@@ -279,10 +280,30 @@ end
...
+maptuple(f, a, b...) = Tuple{f(a), maptuple(f, b...).types...}
+
+# An element type satisfying for all A:
+# broadcast_getindex(
+#     containertype(A),
+#     A, broadcast_indices(A)
+# )::_broadcast_getindex_eltype(A)
+_broadcast_getindex_eltype(A) = _broadcast_getindex_eltype(containertype(A), A)
+_broadcast_getindex_eltype(::ScalarType, T::Type) = Type{T}
+_broadcast_getindex_eltype(::ScalarType, A) = typeof(A)
```
Comment (pabloferz) [inline base/broadcast.jl]: You can subsume this line and the above with `_broadcast_getindex_eltype`(::ScalarType, x) = Core.Typeof(x)`.

File: base/broadcast.jl [hunk 3]
```diff
@@ -279,10 +280,30 @@ end
...
+_broadcast_getindex_eltype(::Any, A) = eltype(A)  # Tuple, Array, etc.
+
+# An element type satisfying for all A:
+# unsafe_get(A)::unsafe_get_eltype(A)
+_unsafe_get_eltype(x::Nullable) = eltype(x)
+_unsafe_get_eltype(T::Type) = Type{T}
+_unsafe_get_eltype(x) = typeof(x)
```
Comment (pabloferz) [inline base/broadcast.jl]: Same as above.

**Comment by pabloferz** [comment]
I like that this uses the `containertype` trait to properly sync what is treated as scalar with the computing of the `eltype` :+1:

File: base/broadcast.jl (hunk 2 above)
Comment (pabloferz) [inline base/broadcast.jl]: Also, how about `_broadcast_container_eltype` instead of `_broadcast_getindex_eltype`?

File: base/broadcast.jl (hunk 3 above)
Comment (pabloferz) [inline base/broadcast.jl]: Is `_unsafe_get_eltype` used anywhere else besides `broadcast`? If not, I think you can fuse it with the `_broadcast_getindex_eltype`.

File: base/broadcast.jl (hunk 2 above)
Comment (Sacha0) [inline base/broadcast.jl]: Guessing @TotalVerb chose `_broadcast_getindex_eltype` to make clear that this function returns the type of the atoms returned by `broadcast_ getindex`?

File: base/broadcast.jl (hunk 3 above)
Comment (Sacha0) [inline base/broadcast.jl]: From another perspective, the explicit connection between `_unsafe_get_eltype` and `unsafe_get` (along with the great comments!) made `_unsafe_get_eltype`'s purpose clear immediately :). Fused into `_broadcast_getindex_eltype`, these methods' purpose might not have been as easy to grasp.

File: base/broadcast.jl [hunk 4]
```diff
@@ -279,10 +280,30 @@ end
     return C
 end
 
-eltypestuple(a) = (Base.@_pure_meta; Tuple{eltype(a)})
-eltypestuple(T::Type) = (Base.@_pure_meta; Tuple{Type{T}})
-eltypestuple(a, b...) = (Base.@_pure_meta; Tuple{eltypestuple(a).types..., eltypestuple(b...).types...})
-_broadcast_eltype(f, A, Bs...) = Base._return_type(f, eltypestuple(A, Bs...))
+maptuple(f) = Tuple{}
+maptuple(f, a, b...) = Tuple{f(a), maptuple(f, b...).types...}
```
Comment (Sacha0) [inline base/broadcast.jl]: IIUC, this function returns a tuple type collecting the values returned when the passed function is applied to each of the trailing arguments of this function (but only where you can construct a tuple type including those values, e.g. not where the passed function returns strings). Perhaps a slightly more descriptive / narrower name would be better, e.g. `maptotupletype` or `maptoTuple`? (I imagine you can come up with something better than those suggestions.) Or do I misunderstand, and `maptuple` somehow better describes this function?

File: base/sparse/higherorderfns.jl (hunk 1 above)
Comment (TotalVerb) [inline base/sparse/higherorderfns.jl]: Yes, good catch. Thanks

**Comment by TotalVerb** [comment]
Appveyor 32 bit failure seems unrelated.

File: base/broadcast.jl (hunk 2 above)
Comment (TotalVerb) [inline base/broadcast.jl]: Yes, that's why.

**Comment by Sacha0** [comment]
AV timeout appears unrelated?

**Review by Sacha0 — APPROVED** [review]
@TotalVerb, this looks great! @pabloferz, thoughts at this stage? Best!

**Comment by pabloferz** [comment]
@nanosoldier `runbenchmarks(["broadcast", "nullable"], vs=":master")`

**Comment by pabloferz** [comment]
Let's try again

@nanosoldier runbenchmarks("broadcast" || "nullable", vs = ":master")

**Comment by pabloferz** [comment]
Third is the charm?

@nanosoldier `runbenchmarks("broadcast" || "nullable", vs = ":master")`

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/21328#issuecomment-293329531) has completed - no performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/b2aeb8724becebbdf19871f00e5ad721192e9e02/dff4513_vs_decdf65/report.md). cc @jrevels

**Review by p-zubieta — APPROVED** [review]

**Comment by timholy** [comment]
This is nice. For any packages that broke with this, consider eliminating any call to `eltypestuple` and see whether the new mechanisms work without override (https://github.com/JuliaGraphics/ColorVectorSpace.jl/pull/74).

**Comment by TotalVerb** [comment]
Sorry! Should have checked if `eltypestuple` was used. (This PR fixes the bug that made the workaround in ColorVectorSpace necessary, so hopefully no other uses of `eltypestuple` exist except to work around the bug.)

Any new overrides should be to `_broadcast_getindex_eltype` and `_unsafe_get_eltype` if the behaviour of `broadcast_getindex` or `unsafe_get` is specified for a particular custom type. This should be more robust than overriding `eltypestuple`. Perhaps this should be documented.

**Comment by timholy** [comment]
This is a nice step towards #20740. It would be great to document the API. Even if the current API isn't yet optimal for specialization, just working on the documentation is a great way to identify any bits that need attention.

File: base/broadcast.jl [hunk 5]
```diff
@@ -279,10 +280,30 @@ end
     return C
 end
 
-eltypestuple(a) = (Base.@_pure_meta; Tuple{eltype(a)})
-eltypestuple(T::Type) = (Base.@_pure_meta; Tuple{Type{T}})
-eltypestuple(a, b...) = (Base.@_pure_meta; Tuple{eltypestuple(a).types..., eltypestuple(b...).types...})
-_broadcast_eltype(f, A, Bs...) = Base._return_type(f, eltypestuple(A, Bs...))
+maptoTuple(f) = Tuple{}
```
Comment (nalimilan) [inline base/broadcast.jl]: This function name does not follow the style guide.

Comment (Sacha0) [inline base/broadcast.jl]: Would you prefer `maptotupletype`? Best!

Comment (TotalVerb) [inline base/broadcast.jl]: This style is actually quite common, including in the inference code. See `isconstType`, `isType` in inference and `Core.Typeof`.

Comment (nalimilan) [inline base/broadcast.jl]: If that's already used elsewhere, I guess that's OK, but it's annoying that we don't follow our style guide in Base code... `maptotupletype` would have looked more standard.

**Pull request merged**